File: app/src/main/java/de/synyx/android/meeroo/domain/MeetingRoom.java
package de.synyx.android.meeroo.domain;

import org.joda.time.Duration;

import java.util.List;

import static org.joda.time.Duration.standardMinutes;


/**
 * @author  Max Dobler - dobler@synyx.de
 */
public class MeetingRoom {

    private Long calendarId;
    private String name;
    private Agenda agenda = new Agenda();

    public MeetingRoom(long calendarId, String name) {

        this.calendarId = calendarId;
        this.name = name;
    }


    public MeetingRoom() {
    }

    public boolean isAvailable() {

        return !agenda.hasCurrentReservation();
    }


    public Duration getTimeUntilNextMeeting() {

        return agenda != null ? agenda.getTimeUntilNextMeeting() : null;
    }


    public MeetingRoom addReservation(Reservation reservation) {

        agenda.addReservation(reservation);

        return this;
    }


    public Duration getTimeUntilAvailable() {

        return agenda.getTimeUntilAvailable();
    }


    public boolean isNextMeetingInNearFuture() {

        return agenda.isNextReservationInNearFuture();
    }


    public long getCalendarId() {

        return this.calendarId;
    }


    public String getName() {

        return name;
    }


    public RoomAvailability getAvailability() {

        if (!isAvailable()) {
            return RoomAvailability.UNAVAILABLE;
        }

        return isNextMeetingInNearFuture() ? RoomAvailability.RESERVED : RoomAvailability.AVAILABLE;
    }


    public Reservation getCurrentMeeting() {

        return agenda.getCurrentMeeting();
    }


    public Reservation getUpcomingReservation() {

        return agenda.getUpcomingReservation();
    }


    public List<Reservation> getReservations() {

        return agenda.getReservations();
    }


    public Duration getAvailabilityTime() {

        RoomAvailability availability = getAvailability();

        Duration duration = availability == RoomAvailability.UNAVAILABLE ? getTimeUntilAvailable()
                                                                         : getTimeUntilNextMeeting();

        if (duration != null) {
            return duration.plus(standardMinutes(1));
        }

        return null;
    }


    public Reservation getSecondUpcomingReserveration() {

        return agenda.getSecondUpcomingReservation();
    }
}

File: app/src/main/java/de/synyx/android/meeroo/domain/MeetingRoom.kt
package de.synyx.android.meeroo.domain

import de.synyx.android.meeroo.domain.RoomAvailability.AVAILABLE
import de.synyx.android.meeroo.domain.RoomAvailability.RESERVED
import org.joda.time.Duration

import org.joda.time.Duration.standardMinutes


/**
 * @author  Max Dobler - dobler@synyx.de
 */
class MeetingRoom(val calendarId: Long, val name: String) {

    private val agenda = Agenda()

    fun isAvailable(): Boolean = !agenda.hasCurrentReservation()


    fun getTimeUntilNextMeeting(): Duration? = agenda.getTimeUntilNextMeeting()


    fun getTimeUntilAvailable(): Duration = agenda.getTimeUntilAvailable()


    fun isNextMeetingInNearFuture(): Boolean = agenda.isNextReservationInNearFuture()


    fun getAvailability(): RoomAvailability {

        if (!isAvailable()) {
            return RoomAvailability.UNAVAILABLE
        }

        return if (isNextMeetingInNearFuture()) RESERVED else AVAILABLE
    }


    fun getCurrentMeeting(): Reservation? = agenda.getCurrentMeeting()


    fun getUpcomingReservation(): Reservation? = agenda.getUpcomingReservation()


    fun getReservations(): List<Reservation> = agenda.getReservations()


    fun getAvailabilityTime(): Duration? {

        val availability = getAvailability()

        val duration = if (availability == RoomAvailability.UNAVAILABLE)
            getTimeUntilAvailable()
        else
            getTimeUntilNextMeeting()

        return duration?.plus(standardMinutes(1))

    }


    fun getSecondUpcomingReserveration(): Reservation? = agenda.getSecondUpcomingReservation()


    fun addReservation(reservation: Reservation) = apply { agenda.addReservation(reservation) }


}
